hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/client/utils/SparkValueMetadata.java

```diff
@@ -0,0 +1,230 @@
+      return ValueType.INT;
+    } else if (dataType instanceof LongType) {
+      return ValueType.LONG;
+    } else if (dataType instanceof FloatType) {
+      return ValueType.FLOAT;
+    } else if (dataType instanceof DoubleType) {
+      return ValueType.DOUBLE;
+    } else if (dataType instanceof StringType || dataType instanceof CharType || dataType instanceof VarcharType) {
+      return ValueType.STRING;
+    }  else if (dataType instanceof TimestampType) {
+      return ValueType.TIMESTAMP_MICROS;
+    }  else if (dataType instanceof DecimalType) {
+      return ValueType.DECIMAL;
+    } else if (dataType instanceof DateType) {
+      return ValueType.DATE;
+    } else if (dataType instanceof BinaryType) {
+      return ValueType.BYTES;
+    } else if (dataType instanceof ArrayType) {
+      if (root) {
+        return fromDataType(((ArrayType) dataType).elementType(), false);
+      } else {
+        throw new IllegalArgumentException("Array of Array is not supported");
+      }
+    } else if (SparkAdapterSupport$.MODULE$.sparkAdapter().isTimestampNTZType(dataType)) {
+      return ValueType.LOCAL_TIMESTAMP_MICROS;
+    } else {
+      throw new IllegalArgumentException("Unsupported data type: " + dataType);
+    }
+  }
+
+  private static class SparkV1EmptyMetadata extends SparkValueMetadata {
+    private static final SparkV1EmptyMetadata V_1_EMPTY_METADATA = new SparkV1EmptyMetadata();
+    public static SparkV1EmptyMetadata get() {
+      return V_1_EMPTY_METADATA;
+    }
+
+    private SparkV1EmptyMetadata() {
+      super(ValueType.V1, null);
+    }
+
+    @Override
+    public HoodieValueTypeInfo getValueTypeInfo() {
+      return null;
+    }
+  }
+
+  private static class SparkDecimalMetadata extends SparkValueMetadata implements ValueMetadata.DecimalValueMetadata {
+
+    private final int precision;
+    private final int scale;
+
+    protected SparkDecimalMetadata(DecimalType decimalType) {
+      super(ValueType.DECIMAL, decimalType);
+      this.precision = decimalType.precision();
+      this.scale = decimalType.scale();
+    }
+
+    @Override
+    public int getPrecision() {
+      return precision;
+    }
+
+    @Override
+    public int getScale() {
+      return scale;
+    }
+
+    @Override
+    public String getAdditionalInfo() {
+      return DecimalValueMetadata.encodeData(this);
+    }
+  }
+
+  private Comparable convertSparkToJava(Object value, boolean rootLevel) {
+    if (value == null) {
+      return null;
+    }
+
+    if (value instanceof org.apache.spark.sql.catalyst.util.ArrayData) {
+      if (rootLevel) {
+        return new ArrayComparable(Arrays.stream(((org.apache.spark.sql.catalyst.util.ArrayData) value)
+                .toObjectArray(((org.apache.spark.sql.types.ArrayType) dataType).elementType()))
+            .map(v -> convertSparkToJava(v, false)).toArray(Comparable[]::new));
+      } else {
+        throw new IllegalArgumentException("array of arrays not supported");
+      }
+    }
```

Comment:
Is this for both V1 and V2 column stats index?  Should we add `ARRAY` as `ValueType` here?